app/src/main/java/app/musikus/ui/activesession/ActiveSessionViewModel.kt

```diff
@@ -3,416 +3,350 @@
+        scope = viewModelScope,
+        started = SharingStarted.WhileSubscribed(5000),
+        initialValue = emptyList()
+    )
 
-    private val itemsInSelectedFolder = _selectedFolderId.flatMapLatest { selectedFolderId ->
-        libraryUseCases.getSortedItems(Nullable(selectedFolderId))
-    }.stateIn(
+    private val runningLibraryItem = activeSessionUseCases.getRunningItem().stateIn(
         scope = viewModelScope,
         started = SharingStarted.WhileSubscribed(5000),
-        initialValue = listOf()
+        initialValue = null
     )
 
-    private val lastPracticedDates = itemsInSelectedFolder.flatMapLatest { items ->
-        libraryUseCases.getLastPracticedDate(items)
+    // TODO: this is actually a 1:1 mapping still, maybe think about either re-using state
+    //       from useCase or extend UI State functionality
+    private val sessionState = activeSessionUseCases.getTimerState().map { state ->
+        when(state) {
+            SessionTimerState.NOT_STARTED -> ActiveSessionState.NOT_STARTED
+            SessionTimerState.RUNNING -> ActiveSessionState.RUNNING
+            SessionTimerState.PAUSED -> ActiveSessionState.PAUSED
+            SessionTimerState.UNKNOWN -> ActiveSessionState.UNKNOWN // TODO evaluate what to do
+        }
     }.stateIn(
         scope = viewModelScope,
         started = SharingStarted.WhileSubscribed(5000),
-        initialValue = emptyMap()
+        initialValue = ActiveSessionState.UNKNOWN
     )
 
-    private val foldersWithItems = libraryUseCases.getSortedFolders().stateIn(
+    private val libraryFoldersWithItems = libraryUseCases.getSortedFolders().stateIn(
         scope = viewModelScope,
         started = SharingStarted.WhileSubscribed(5000),
-        initialValue = listOf()
+        initialValue = emptyList()
     )
 
-    private val completedSections = activeSessionUseCases.getCompletedSections().stateIn(
+    private val rootItems = libraryUseCases.getSortedItems(Nullable(null)).stateIn(
         scope = viewModelScope,
         started = SharingStarted.WhileSubscribed(5000),
         initialValue = emptyList()
     )
 
-    private val runningLibraryItem = activeSessionUseCases.getRunningItem().stateIn(
+    private val allLibraryItems = combine(
+        libraryFoldersWithItems,
+        rootItems
+    ) { folders, rootItems ->
+        folders.flatMap { it.items } + rootItems
+    }
+
+    @OptIn(ExperimentalCoroutinesApi::class)
+    private val lastPracticedDates = allLibraryItems.flatMapLatest {
+        libraryUseCases.getLastPracticedDate(it)
+    }.stateIn(
         scope = viewModelScope,
         started = SharingStarted.WhileSubscribed(5000),
-        initialValue = null
+        initialValue = emptyMap()
     )
 
 
-    /**
-     *  ---------------- Composing the Ui state --------------------
-     */
+    /** ------------------- Own StateFlow UI states  ------------------------------------------- */
```

Comment:
IMO it would be a good rule for us to never explicitly store _UiState_ in a variable inside the ViewModel and only derive it from private StateFlow.

Reply:
but they are *private* StateFlows, this is basically the lowest-level UI state there is, because they are not derived from UseCases. since all are packed in one StateFlow and will have to be updated together anyways, it makes no sense to define another variable for every property plus another state for wrapping all the properties inside the data class.

There would be no functional benefit but largely more useless boilerplate code, or am I wrong?

Reply:
I still like the idea, that the ViewModel has its _own_ stateFlows, _imported_ stateFlows and maybe _combined_ stateFlows which comprise the ViewModel **state**. And from that state, the UiState is entirely derived even if it is a 1-to-1 mapping of a ViewModel state to UiState. This way we never write to UiState.

Reply:
I agree 👍. That's exactly what I did. I never used the StateFlows from the useCases but created own ones, even though they are just 1:1 mappings (that's why I called them proxies).

So all combine statements compose themselves only from private StateFlows of the ViewModel

Reply:
I guess it's just a naming convention at that point. I would have omitted the "UiState" suffix for these StateFlows.